processor/processor.go

```diff
@@ -20,44 +20,65 @@ package processor
 import (
 	"github.com/santhosh-tekuri/jsonschema"
 
-	"github.com/elastic/apm-server/model"
+	"github.com/elastic/apm-server/decoder"
+	"github.com/elastic/apm-server/model/metadata"
+	"github.com/elastic/apm-server/transform"
+	"github.com/elastic/apm-server/utility"
 	"github.com/elastic/apm-server/validation"
 	"github.com/elastic/beats/libbeat/monitoring"
 )
 
 type Processor interface {
 	Validate(map[string]interface{}) error
-	Decode(map[string]interface{}) (model.Payload, error)
+	Decode(map[string]interface{}) (*metadata.Metadata, []transform.Transformable, error)
 	Name() string
 }
 
-type PayloadDecoder func(map[string]interface{}) (model.Payload, error)
-
-type PayloadProcessor struct {
-	ProcessorName string
-	DecodePayload PayloadDecoder
+type EventsProcessor struct {
+	PluralKeyName string
+	SingularName  string
+	EventDecoder  decoder.EventDecoder
 	PayloadSchema *jsonschema.Schema
 	DecodingCount *monitoring.Int
 	DecodingError *monitoring.Int
 	ValidateCount *monitoring.Int
 	ValidateError *monitoring.Int
 }
 
-func (p *PayloadProcessor) Name() string {
-	return p.ProcessorName
+func (e *EventsProcessor) Name() string {
+	return e.SingularName
```

Comment:
This is only used for the tests. Let me know if you have better ideas: https://github.com/elastic/apm-server/blob/master/script/output_data/output_data.go#L46

Reply:
How about you simply use the pluralized name and remove the `s` in the test?